DataStructures/BinarySearchTree/BinarySearchTree.cs

```diff
@@ -0,0 +1,421 @@
+    /// <item>All nodes in the Left subtree contain keys that are less than the node's key.</item>
+    /// <item>All nodes in the Right subtree contain keys that are greater than the node's key.</item>
+    /// </list>
+    /// A BST will have an average complexity of O(log n) for insertion, removal, and lookup operations.
+    /// </remarks>
+    /// <typeparam name="TKey">Type of key for the BST. Keys must implement IComparable.</typeparam>
+    public class BinarySearchTree<TKey> where TKey : IComparable<TKey>
+    {
+        /// <summary>
+        /// Private class member to hold the root of the BST.
+        /// </summary>
+        private BinarySearchTreeNode<TKey>? root;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="BinarySearchTree{TKey}"/> class.
+        /// </summary>
+        public BinarySearchTree()
+        {
+            root = null;
+            Count = 0;
+        }
+
+        /// <summary>
+        /// Gets the number nodes currently in the BST.
+        /// </summary>
+        public int Count { get; private set; }
+
+        /// <summary>
+        /// Insert a key into the BST.
+        /// </summary>
+        /// <param name="key">The key to insert.</param>
+        /// <exception cref="ArgumentNullException">
+        /// Thrown if key is null.
+        /// </exception>
+        /// <exception cref="ArgumentException">
+        /// Thrown if key is already in BST.
+        /// </exception>
+        public void Add(TKey key)
+        {
+            if (key == null)
+            {
+                throw new ArgumentNullException("Key cannot be null!");
+            }
+
+            if (root == null)
+            {
+                root = new BinarySearchTreeNode<TKey>(key);
+            }
+            else
+            {
+                Add(root, key);
+            }
+
+            Count++;
+        }
+
+        /// <summary>
+        /// Insert multiple keys into the BST.
+        /// Keys are inserted in the order they appear in the sequence.
+        /// </summary>
+        /// <param name="keys">Sequence of keys to insert.</param>
+        /// <exception cref="ArgumentException">
+        /// Thrown if a key in the sequence is already in BST.
+        /// </exception>
+        public void AddRange(IEnumerable<TKey> keys)
+        {
+            foreach (TKey key in keys)
+            {
+                Add(key);
+            }
+        }
+
+        /// <summary>
+        /// Find a node with the specified key.
+        /// </summary>
+        /// <param name="key">The key to search for.</param>
+        /// <returns>The node with the specified key if it exists, otherwise a default value is returned.</returns>
+        /// <exception cref="ArgumentNullException">
+        /// Thrown if key is null.
+        /// </exception>
+        public BinarySearchTreeNode<TKey>? Search(TKey key)
+        {
+            if (key == null)
```

Comment:
Codacy found an issue: [Use a comparison to 'default(TKey)' instead or add a constraint to 'TKey' so that it can't be a value type.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6277689)